packetbeat/protos/tls/parse.go

```diff
@@ -0,0 +1,506 @@
+	m["cipher"] = s.cipher.String()
+	m["compression_method"] = s.compMethod.String()
+	m["extensions"] = s.extensions
+	return m
+}
+
+func (parser *parser) parse(buf *streambuf.Buffer, conn *tlsConnectionData, plugin *tlsPlugin) (bool, bool) {
+
+	if conn.handshakeCompleted {
+		// ignore encrypted traffic
+		buf.Advance(buf.Len())
+		return true, true
+	}
+
+	var header recordHeader
+	for buf.Avail(recordHeaderSize) {
+
+		if err := header.read(buf); err != nil {
+			logp.Warn("tls", "internal buffer error: %v", err)
+			return false, false
+		}
+
+		if !header.isValid() {
+			// ditch this stream
+			return false, false
+		}
+
+		limit := recordHeaderSize + int(header.length)
+		if !buf.Avail(limit) {
+			// wait for complete record
+			return true, false
+		}
+
+		switch header.recordType {
+		case recordTypeChangeCipherSpec: // single message of size 1 (byte 1)
+			conn.handshakeCompleted = true
+			plugin.sendEvent(conn)
+			if isDebug {
+				logp.Debug("tls", "transitioned to handshakeCompleted mode")
+			}
+			// discard remaining data for this stream (encrypted)
+			buf.Advance(buf.Len())
+
+		case recordTypeHandshake:
+			if isDebug {
+				logp.Debug("tls", "got handshake record of size %d", header.length)
+			}
+			if !parser.parseHandshake(buf, int(header.length), conn) {
+				return false, false
+			}
+
+		case recordTypeAlert:
+			if err := parser.parseAlert(&header, buf); err != nil {
+				logp.Warn("Error parsing alert message: %v", err)
+				return false, false
+			}
+
+		case recordTypeApplicationData:
+			// TODO: Request / Response analytics
+			if isDebug {
+				logp.Debug("tls", "ignoring application data of size %d", header.length)
+			}
+
+		default:
+			if isDebug {
+				logp.Debug("tls", "ignoring record type %d", header.recordType)
+			}
+		}
+
+		buf.Advance(limit)
+	}
+
+	if buf.Len() == 0 {
+		return true, true
+	}
+	return true, false
+}
+
+func (parser *parser) parseHandshake(buf *streambuf.Buffer, length int, conn *tlsConnectionData) bool {
+	// TODO: parse in-place if message in packet buffer is complete
+	if err := parser.handshakeBuf.Append(buf.Bytes()[recordHeaderSize : recordHeaderSize+length]); err != nil {
+		logp.Warn("failed appending to handshake buffer: %v", err)
+		return false
+	}
+	for parser.handshakeBuf.Avail(handshakeHeaderSize) {
+		// type
+		var header handshakeHeader
+		err := header.read(&parser.handshakeBuf)
+		if err != nil {
+			logp.Warn("failed reading handshake: %v", err)
+			return false
+		}
+		limit := handshakeHeaderSize + header.length
+		if limit > parser.handshakeBuf.Len() {
+			break
+		}
+		parser.processHandshake(conn, header.handshakeType,
+			bufferView{&parser.handshakeBuf, handshakeHeaderSize, limit})
+		parser.handshakeBuf.Advance(limit)
+	}
+	if parser.handshakeBuf.Len() == 0 {
+		parser.handshakeBuf.Reset()
+	}
+	return true
+}
+
+func (parser *parser) processHandshake(conn *tlsConnectionData, handshakeType handshakeType, reader bufferView) bool {
```

Comment:
This looks like a task for the message 'correlator'. Note, depending on deployment (e.g. port forwarding), packetbeat might potentially see packets out of order.